gobblin-utility/src/main/java/gobblin/util/PullFileLoader.java

```diff
@@ -0,0 +1,257 @@
+   * @param sysProps A {@link Config} used as fallback.
+   * @param loadGlobalProperties if true, will also load at most one *.properties file per directory from the
+   *          {@link #rootDirectory} to the pull file {@link Path}.
+   * @return The loaded {@link Config}.
+   * @throws IOException
+   */
+  public Config loadPullFile(Path path, Config sysProps, boolean loadGlobalProperties) throws IOException {
+    Config fallback = loadGlobalProperties ? loadAncestorGlobalConfigs(path, sysProps) : sysProps;
+
+    if (this.javaPropsPullFileFilter.accept(path)) {
+      return loadJavaPropsWithFallback(path, fallback).resolve();
+    } else if (this.hoconPullFileFilter.accept(path)) {
+      return loadHoconConfigAtPath(path).withFallback(fallback).resolve();
+    } else {
+      throw new IOException(String.format("Cannot load pull file %s due to unrecognized extension.", path));
+    }
+  }
+
+  /**
+   * Find and load all pull files under a base {@link Path} recursively.
+   * @param path base {@link Path} where pull files should be found recursively.
+   * @param sysProps A {@link Config} used as fallback.
+   * @param loadGlobalProperties if true, will also load at most one *.properties file per directory from the
+   *          {@link #rootDirectory} to the pull file {@link Path} for each pull file.
+   * @return The loaded {@link Config}s.
+   */
+  public Collection<Config> loadPullFilesRecursively(Path path, Config sysProps, boolean loadGlobalProperties) {
+    try {
+      Config fallback = sysProps;
+      if (loadGlobalProperties && PathUtils.isAncestor(this.rootDirectory, path.getParent())) {
+        fallback = loadAncestorGlobalConfigs(path.getParent(), fallback);
+      }
+      return loadPullFilesRecursivelyHelper(path, fallback, loadGlobalProperties);
+    } catch (IOException ioe) {
+      return Lists.newArrayList();
+    }
+  }
+
+  private Collection<Config> loadPullFilesRecursivelyHelper(Path path, Config fallback, boolean loadGlobalProperties) {
+    List<Config> pullFiles = Lists.newArrayList();
+
+    try {
+      if (loadGlobalProperties) {
+        fallback = findAndLoadGlobalConfigInDirectory(path, fallback);
+      }
+
+      FileStatus[] statuses = this.fs.listStatus(path);
+      if (statuses == null) {
+        log.error("Path does not exist: " + path);
+        return pullFiles;
+      }
+
+      for (FileStatus status : statuses) {
+        if (status.isDirectory()) {
+          pullFiles.addAll(loadPullFilesRecursivelyHelper(status.getPath(), fallback, loadGlobalProperties));
+        } else if (this.javaPropsPullFileFilter.accept(status.getPath())) {
+          pullFiles.add(loadJavaPropsWithFallback(status.getPath(), fallback).resolve());
+        } else if (this.hoconPullFileFilter.accept(status.getPath())) {
+          pullFiles.add(loadHoconConfigAtPath(status.getPath()).withFallback(fallback).resolve());
+        }
+      }
+
+      return pullFiles;
+    } catch (IOException ioe) {
```

Comment:
Hmm, you're right, and that's a problem. The idea is that if a subdirectory fails to load, pull files in other directories will still be loaded. I'll fix that.